lucene/sandbox/src/java/org/apache/lucene/search/LargeNumHitsTopDocsCollector.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.search;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+import org.apache.lucene.index.LeafReaderContext;
+
+import static org.apache.lucene.search.TopDocsCollector.EMPTY_TOPDOCS;
+
+/**
+ * Optimized collector for large number of hits.
+ * The collector maintains an ArrayList of hits until it accumulates
+ * the requested number of hits. Post that, it builds a Priority Queue
+ * and starts filtering further hits based on the minimum competitive
+ * score.
+ */
+public final class LargeNumHitsTopDocsCollector implements Collector {
+  private final int requestedHitCount;
+  private List<ScoreDoc> hits = new ArrayList<>();
+  // package private for testing
+  HitQueue pq;
+  ScoreDoc pqTop;
+  int totalHits;
+
+  public LargeNumHitsTopDocsCollector(int requestedHitCount) {
+    this.requestedHitCount = requestedHitCount;
+    this.totalHits = 0;
+  }
+
+  // We always return COMPLETE since this collector should ideally
+  // be used only with large number of hits case
+  @Override
+  public ScoreMode scoreMode() {
+    return ScoreMode.COMPLETE;
+  }
+
+  @Override
+  public LeafCollector getLeafCollector(LeafReaderContext context) {
+    final int docBase = context.docBase;
+    return new TopScoreDocCollector.ScorerLeafCollector() {
+
+      @Override
+      public void setScorer(Scorable scorer) throws IOException {
+        super.setScorer(scorer);
+      }
+
+      @Override
+      public void collect(int doc) throws IOException {
+        float score = scorer.score();
+
+        // This collector relies on the fact that scorers produce positive values:
+        assert score >= 0; // NOTE: false for NaN
+
+        if (totalHits < requestedHitCount) {
+          hits.add(new ScoreDoc(doc, score));
```

Comment:
OK I dug the test failure, it was due to the fact that you are passing `doc` here instead of `docBase + doc`. I'm curious why you couldn't find it, it happened pretty often for me.